src/primitives/polynomial.rs

```diff
@@ -0,0 +1,277 @@
+// SPDX-License-Identifier: MIT
+
+//! Polynomials over Finite Fields
+
+#[cfg(all(feature = "alloc", not(feature = "std"), not(test)))]
+use alloc::vec::Vec;
+use core::{iter, ops};
+
+use super::Field;
+
+/// A polynomial over some field.
+#[derive(PartialEq, Eq, Clone, Debug, Hash)]
+pub struct Polynomial<F> {
+    inner: Vec<F>,
+}
+
+impl<F> Polynomial<F> {
+    /// Constructor for a polynomial from a vector of coefficients.
+    ///
+    /// These coefficients are in "little endian" order. That is, the ith
+    /// coefficient is the one multiplied by x^i.
+    pub fn new(f: Vec<F>) -> Self { Self { inner: f } }
+}
+
+impl<F: Field> Polynomial<F> {
+    /// The degree of the polynomial.
+    ///
+    /// For constants it will return zero, including for the constant zero.
+    pub fn degree(&self) -> usize {
+        self.inner.len() - 1 - self.inner.iter().rev().take_while(|el| **el == F::ZERO).count()
```

Comment:
Maybe an intermediate variable?

```rust
let trailing_zero_terms = self.inner.iter().rev().take_while(|el| **el == F::ZERO).count();
// A degree-N polynomial has N+1 terms
self.inner.len() - trailing_zero_terms - 1
```

Q: Does this go out of bounds when all terms are zeros?

Reply:
It will panic if there is an empty array, yes, but this should never happen. (The zero polynomial has one term which should be `Fe32(0)`.) I can add a debug_assert to make this explicit.

Reply:
Pushed to add a debug assertion and named variables.